### PR TITLE
JAX ecosystem: `jaxlib` should never be in `propagatedBuildInputs`

### LINKED ISSUE #156767: python3Packages.jaxlib: NVIDIA GPU not detected correctly
### Describe the bug
The jax/jaxlib python packages are not detecting NVIDIA GPUs on EC2 instances running NixOS. This is due to other packages including `jaxlib` in their `propagatedBuildInputs` instead of `checkInputs`. I'll follow up with a PR to fix all of these instances.

### Steps To Reproduce
1. Run a g4dn.xlarge EC2 instance including the following in its `/etc/nixos/configuration.nix`
```nix
  # NVIDIA drivers, etc.
  nixpkgs.config.allowUnfree = true;
  services.xserver.videoDrivers = [ "nvidia" ];
  hardware.opengl.enable = true;
```
2. Set up a `shell.nix` file as follows
```nix
let
  # Last updated: 1/17/2022. Check for new commits at status.nixos.org.
  pkgs = import (fetchTarball ("https://github.com/NixOS/nixpkgs/archive/bc59ba15b64d0a0ee1d1764f18b4f3480d2c3e5a.tar.gz")) { };
in
pkgs.mkShell {
  buildInputs = with pkgs; [
    python3
    python3Packages.flax
    python3Packages.jax
    (python3Packages.jaxlib.override { cudaSupport = true; })
  ];
}
```
3. Note
```
❯ nix-shell shell.nix --run "python -c 'import jax; print(jax.devices())'"
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
[<jaxlib.xla_extension.Device object at 0x7f6a26b50270>]
```

OTOH, removing `flax` from shell.nix makes everything work just fine. This appears to be due to packages including `jaxlib` in their `propagatedBuildInputs` which creates multiple jaxlib editions in `$PYTHONPATH`:
```
❯ nix-shell shell.nix --run "echo \$PYTHONPATH | tr ':' '\n' | grep jaxlib"
/nix/store/2b9adfxqq0ix45q0qslhwwrl86z5jw97-python3.9-jaxlib-0.1.75/lib/python3.9/site-packages
/nix/store/qrddbps5cq62n2d0fg05dm73ykxzsyv3-python3.9-jaxlib-0.1.75/lib/python3.9/site-packages
```
This means the CPU version of jaxlib is clobbering the GPU-version. We need to remove `jaxlib` from the `propagatedBuildInputs` from all packages so that users have the choice of which version of jaxlib to use.

### Expected behavior
GPU to be automatically detected when the correct drivers are installed/configured.

### Screenshots
n/a

### Additional context
I initially suspected that this was related to the transition to a source-build of jaxlib in https://github.com/NixOS/nixpkgs/pull/151909, but that appears to be unrelated.

### Notify maintainers

@samuela @ndl 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.91, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.16`
 - channels(root): `"nixos-21.11.335376.610d4ea2750, nix-ld"`
 - channels(skainswo): `"home-manager, nixpkgs-unstable-22.05pre343295.adf7f03d3bf"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by samuela**
###### Motivation for this change
Fix https://github.com/NixOS/nixpkgs/issues/156767.

There are multiple possible "versions" of the jaxlib package: CPU-only, CUDA-enabled, and TPU-enabled. Currently the jaxlib package defaults to the smallest (and most FOSS) version which is CPU-only. CUDA support can be enabled by end users via the `cudaSupport = true` option. 

This presents an issue when using packages that depend on jax/jaxlib however. If the packages depend directly on jaxlib by adding them to `propagatedBuildInputs` then the end-user is stuck with the CPU-only version of jaxlib, even though the CUDA version may be desired. Using overlays is one possible workaround but unfortunately that requires every end-user to recompile the entire JAX ecosystem locally whenever they want to use a non-CPU jaxlib.

This PR removes jaxlib from any and all `propagatedBuildInputs` and patches packages that require jaxlib as a `setup.py` dependency. AFAIU this is in line with the JAX authors' intention to separate the jax and jaxlib packages. In fact, jax itself does not even have a `setup.py` dependency on jaxlib.

For future debugging purposes, an easy way to test if you have conflicting jaxlib versions is 
```
echo $PYTHONPATH | tr ':' '\n' | grep jaxlib
```
This command should return at most one entry at all times. I've verified that the changes in this PR do in fact resolve this issue based on the shell.nix used in https://github.com/NixOS/nixpkgs/issues/156767.

Assuming we reach consensus that this is the right path forward, I will document this practice on the [wiki page](https://nixos.wiki/wiki/JAX).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 156808` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages failed to build:</summary>
  <ul>
    <li>python310Packages.dm-haiku</li>
    <li>python310Packages.elegy</li>
    <li>python310Packages.flax</li>
    <li>python310Packages.objax</li>
    <li>python310Packages.optax</li>
    <li>python310Packages.treex</li>
    <li>python39Packages.objax</li>
  </ul>
</details>
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>python310Packages.chex</li>
    <li>python310Packages.jmp</li>
    <li>python310Packages.treeo</li>
    <li>python39Packages.chex</li>
    <li>python39Packages.dm-haiku</li>
    <li>python39Packages.elegy</li>
    <li>python39Packages.flax</li>
    <li>python39Packages.jmp</li>
    <li>python39Packages.optax</li>
    <li>python39Packages.treeo</li>
    <li>python39Packages.treex</li>
  </ul>
</details>

**Comment by samuela** [comment]
I'm not worried about python310 failures as that entire package set is mostly broken atm. `python39Packages.objax` builds successfully outside of nixpkgs-review for me. No idea why nixpkgs-review keeps doing this. 

cc @Mic92 for any thoughts?

File: pkgs/development/python-modules/chex/default.nix
Comment (FRidh) [inline pkgs/development/python-modules/chex/default.nix:31]: postPatch

File: pkgs/development/python-modules/objax/default.nix
Comment (FRidh) [inline pkgs/development/python-modules/objax/default.nix:26]: postPatch

File: pkgs/development/python-modules/optax/default.nix
Comment (FRidh) [inline pkgs/development/python-modules/optax/default.nix:25]: postPatch

**Review by FRidh — CHANGES_REQUESTED** [review]
Please report these issues also with the upstream users of jaxlib so they can fix their requirements.

**Comment by ndl** [comment]
For the packages that explicitly depend on `jaxlib` in their `setup.py`: IIUC the alternative to patches is to move `jaxlib` from `propagatedBuildInputs` to `buildInputs`? That's seems sufficient to satisfy `setup.py` requirements, have imports checks and tests pass for the package itself - while still not propagating `jaxlib` to the dependent packages?

**Comment by samuela** [comment]
@ndl If that's the case it would certainly make my life easier, and does seem cleaner on the nix side of things. I'll give it a shot and see if I can get it to work!

**Comment by samuela** [comment]
I tested the `checkInputs` solution as suggested by @ndl and sure enough it works! This PR is ready to review/merge.

**Review by ndl — APPROVED** [review]
LG, thanks for fixing this!

**Comment by samuela** [comment]
Thanks @ndl! I'll merge tomorrow unless anyone objects

**Comment by dotlambda** [comment]
This is not the solution to go with! If people want to use a different version of jaxlib they can resort to python's package overrides.

**Comment by samuela** [comment]
@dotlambda there are two main problems with package overrides:
* overriden packages do not get built, tested, and cached by hydra or cache.nixos.org
* the JAX authors separated jax and jaxlib for just this reason that user's have the freedom to select which jaxlib version suits them best

**Comment by dotlambda** [comment]
> * overriden packages do not get built, tested, and cached by hydra or cache.nixos.org

neither does the unfree jaxlibWithCuda



> * the JAX authors intentionally separated jax and jaxlib for just this reason so that users have the freedom to select which jaxlib version suits them best

And people easily get that freedom:
```nix
let
  python = python3.override {
    packageOverrides = self: super: {
      jaxlib = self.jaxlibWithCuda;
    };
  };
in python.pkgs.elegy
```
I suggest that this is put in the wiki instead of the current solution.

**Comment by samuela** [comment]
> neither does the unfree jaxlibWithCuda

True, although this makes things much easier for end users maintaining their own CI infrastructure. For example, [nixpkgs-upkeep](https://github.com/samuela/nixpkgs-upkeep/) builds `jaxlibWithCuda` automatically and pushes it to [my public cachix cache](https://app.cachix.org/cache/ploop#pull).

> And people easily get that freedom:

Overlays are a constant source of confusion for users (see eg https://discourse.nixos.org/t/what-are-overlays/14680), and are not nearly as ergonomic as simply `jaxlibWithCuda`. At minimum, any solution here must work for users writing one-off scripts with nix-shell shebang lines.

By the way, there is extensive precedent for this pattern in nixpkgs already: `tensorflowWithCuda`, `katagoWithCuda`, `librealsenseWithCuda`, `TheanoWithCuda`, `colmapWithCuda`, etc.

**Comment by dotlambda** [comment]
cc @jonringer @FRidh @mweinelt

**Comment by ndl** [comment]
FWIW I've actually used the overlay / override solution in the past for a locally-maintained CUDA version of `jaxlib` but I like @samuela approach better for a number of reasons:
* No need for (not fully trivial) rebuilds of all JAX-related packages when switching `jaxlib` variants.
* Very clear and intuitive composability: just specify which `jaxlib` version you want in your `nix-shell` or derivation and you're done (compare that to https://github.com/NixOS/nixpkgs/issues/44426 and some of the solutions proposed).
* Matches the upstream approach - `jax` [doesn't depend on `jaxlib`](https://github.com/google/jax/blob/main/setup.py) in its core set of dependencies.
* If someone prefers to have `jax` bundled with `jaxlib` - then with the current approach it's pretty easy to produce such a derivation by composing `jax` and `jaxlib` of one's choice together, and then for all intents and purposes it will behave as if these were bundled originally, whereas doing the opposite (= unbundling them if they were originally bundled) is more difficult.

@dotlambda are there any particular issues that you see with this approach that outweigh the advantages listed above?

**Comment by dotlambda** [comment]
> are there any particular issues that you see with this approach that outweigh the advantages listed above?

Any package that depends on a package that depends on a package that ... that uses jax needs to have jaxlib in its buildInputs.

> * Matches the upstream approach - `jax` [doesn't depend on `jaxlib`](https://github.com/google/jax/blob/main/setup.py) in its core set of dependencies.

If you want to be able to switch out packages without rebuilding its reverse dependencies (like upstream wants you to be able to) then you should use something other than Nix.

**Comment by samuela** [comment]
> If you want to be able to switch out packages without rebuilding its reverse dependencies (like upstream wants you to be able to) then you should use something other than Nix.

I don't feel like this is a practical solution to any user's problems. I feel as though the message here has devolved a bit into "no, not that way" without offering an alternative solution that actually meets our needs. As maintainers (and users!) of the JAX ecosystem, @ndl and I have found a solution that works very very well for us. I'm open to entertaining alternative designs to meet our specific needs, but we need concrete ideas not just "don't do it".

**Comment by ndl** [comment]
> Any package that depends on a package that depends on a package that ... that uses jax needs to have jaxlib in its buildInputs.

Right now I think the split is actually closer to 50 / 50: 4 packages (`chex`, `dm-haiku`, `jmp`, `treeo`) *don't* have `jaxlib` dependency in their requirements (and, correspondingly, don't need it in `buildInputs`) while 5 packages (`elegy`, `flax`, `objax`, `optax`, `treex`) do. Out of these 5, I believe exactly zero actually *need* that dependency at build time => the case can be made for raising the corresponding issues with the upstream to remove this dependency.

**Comment by dotlambda** [comment]
> we need concrete ideas not just "don't do it".

What about using `packageOverrides` in your `shell.nix` wouldn't work?
If you want cached packages then you could put all packages built with jaxlibWithCuda on cachix.


> I believe exactly zero actually _need_ that dependency at build time => the case can be made for raising the corresponding issues with the upstream to remove this dependency.

`propagatedBuildInputs` is for run time dependencies, not build time

**Comment by dotlambda** [comment]
This is a debate between purity and convenience and I'd go with purity because I don't wanna be hit by bad decisions later on.

**Comment by samuela** [comment]
> What about using packageOverrides in your shell.nix wouldn't work?

`packageOverrides` is too complex. Esp. for what is effectively the normative use-case of this particular piece of software; the vast majority of people want to use `jaxlibWithCuda`, not `jaxlib`. The average ML researcher is not excited about building/packaging/installing software. They want something that "just works". It can't be more complicated than the `pip install` that they're accustomed to.

`packageOverrides` also presents an issue when writing scripts with nix-shell shebang lines. Even if it's technically possible, it would certainly be messy.

> If you want cached packages then you could put all packages built with jaxlibWithCuda on cachix.

This means every package in the JAX ecosystem ever. It's just too much. Even forgetting about the compute costs, it would make CI much messier.  Especially considering that upstream was very intentional in making the design decision to avoid this problem.

Bear in mind that we've only been considering the CPU and GPU versions of jaxlib, but there's a TPU version as well that's not even packaged yet! As new specialized DL hardware is developed other versions will likely be created as well.

> This is a debate between purity and convenience and I'd go with purity because I don't wanna be hit by bad decisions later on.

As with any software project, it's always possible that we may run into unforeseen complications in the future. And if/when that happens, I think it would be a more appropriate time to have this conversation. The current design works well for the road bumps that we've hit up to this point. And as @ndl mentioned, the modularity of our current design is strictly more flexible than putting `jaxlib` in `propagatedBuildInputs`, meaning that the transition in that direction would be much easier than the other way around.

**Pull request merged**
